core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -536,26 +536,31 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
         return new Engine.Index(uid, doc, version, versionType, origin, startTime, autoGeneratedIdTimestamp, isRetry);
     }
 
-    public void index(Engine.Index index) {
+    public Engine.IndexResult index(final Engine.Index index) {
         ensureWriteAllowed(index);
         Engine engine = getEngine();
-        index(engine, index);
+        return index(engine, index);
     }
 
-    private void index(Engine engine, Engine.Index index) {
+    private Engine.IndexResult index(final Engine engine, final Engine.Index index) {
         active.set(true);
-        index = indexingOperationListeners.preIndex(index);
```

Comment:
I agree with this, but it's a different change and is not related to this one? shall we do it another PR?